### DISCUSSION TITLE
Defer @@unique constraint until end of transaction

### DISCUSSION THREAD

**Discussion by vphoebe** [prisma/prisma#8789] [Q&A] [answered]
I'm using prisma and postgres to organize a music queue for a Discord bot, what basically amounts to a playlist. My model is:
```prisma
model Track {
  guildId      String
  queueIndex   Int
  service      String
  title        String
  url          String
  user         String
  thumbnailUrl String
  lengthInSec  Int
  id           String @id @default(cuid())

  @@unique([guildId, queueIndex], name: "queuePosition")
}
``` 
During insert operations (eg. add a track to the middle of the queue), the `queueIndex` must be incremented or decremented based on the operation, but these fail on the unique constraint because another `Track` already has the same `queueIndex`. 

Using this guide, I'd like to be able to "defer" the unique constraint until the end of the transaction, therefore allowing "shuffling" between the indices. https://medium.com/the-missing-bit/keeping-an-ordered-collection-in-postgresql-9da0348c4bbe

There doesn't seem to be a native way to do this with prisma, but I can replicate if I
1. Remove the ``@@unique`` constraint from the schema.
2. Run custom SQL to add the constraint manually, with the deferral allowed. 
```sql
alter table "Track" add constraint "queuePosition"
  unique("guildId", "queueIndex")
  deferrable initially deferred;
```
3. Run `npx prisma db pull` to replicate it back into the schema. Prisma still adds the exact same `@@unique` string but the database now allows the transactions because the unique constraint isn't enforced till after the transaction.

Is there a way to do this natively with the prisma schema that I am missing? Thanks for your help!

**Comment by ryands17** [comment, marked as answer]
@vphoebe 👋 

Yes this is only possible via Migrate at the moment. It would be great if you could add a feature request for this [here](https://github.com/prisma/prisma/issues/new?assignees=&labels=&template=feature_request.md&title=) so that we can look into this 🙂

**Reply by vphoebe** [reply]
Thanks! Feature request opened here: https://github.com/prisma/prisma/issues/8807

When you mentioned it's possible via Migrate, are you alluding to custom migrations like this? https://www.prisma.io/docs/guides/database/developing-with-prisma-migrate/customizing-migrations I wasn't aware of this originally, so maybe this would help as a workaround.

**Reply by ryands17** [reply]
Yes exactly this 🙂

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
